overlord/managers_test.go

```diff
@@ -6595,3 +6575,441 @@ volumes:
+	tError := st.NewTask("error-trigger", "kernel failed")
+	for _, ts := range tasksets {
+		// skip the taskset of UpdateMany that does the
+		// check-refresh, see tsWithoutReRefresh for details
+		tasks := ts.Tasks()
+		if tasks[0].Kind() == "check-rerefresh" {
+			continue
+		}
+
+		snapsup, err := snapstate.TaskSnapSetup(tasks[0])
+		c.Assert(err, IsNil)
+		// trigger an error as last operation of gadget refresh
+		if snapsup.SnapName() == "pi-kernel" {
+			last := tasks[len(tasks)-1]
+			tError.WaitFor(last)
+			// XXX: or just use "snap-setup" here?
+			tError.Set("snap-setup-task", tasks[0].ID())
+			ts.AddTask(tError)
+			// must be in the same lane as the gadget upate
+			lanes := last.Lanes()
+			c.Assert(lanes, HasLen, 1)
+			tError.JoinLane(lanes[0])
+		}
+
+		chg.AddAll(ts)
+	}
+
+	st.Unlock()
+	err = ms.o.Settle(settleTimeout)
+	st.Lock()
+	c.Assert(err, IsNil)
+	c.Check(chg.Err(), IsNil)
+
+	// At this point the gadget and kernel are updated and the kernel
+	// required a restart. Check that *before* this restart the DTB
+	// files from the kernel are in place.
+	c.Check(filepath.Join(dirs.GlobalRootDir, "/run/mnt/", structureName, "bcm2710-rpi-2-b.dtb"), testutil.FileContains, "bcm2710-rpi-2-b.dtb rev2-from-kernel")
+	c.Check(filepath.Join(dirs.GlobalRootDir, "/run/mnt/", structureName, "bcm2710-rpi-3-b.dtb"), testutil.FileContains, "bcm2710-rpi-3-b.dtb rev2-from-kernel")
+	c.Check(filepath.Join(dirs.GlobalRootDir, "/run/mnt/", structureName, "overlays/uart0.dtbo"), testutil.FileContains, "uart0.dtbo rev2-from-kernel")
+	//  gadget content is not updated because there is no edition update
+	c.Check(filepath.Join(dirs.GlobalRootDir, "/run/mnt/", structureName, "start.elf"), testutil.FileContains, "start.elf rev1")
+
+	// pretend we restarted
+	ms.mockSuccessfulReboot(c, ms.bloader, []snap.Type{snap.TypeKernel})
+
+	st.Unlock()
+	err = ms.o.Settle(settleTimeout)
+	st.Lock()
+	c.Assert(err, IsNil)
+	c.Check(chg.Err(), ErrorMatches, `cannot perform the following tasks:\n- kernel failed.*`)
+
+	// Ensure that tasks states are what we expect
+	taskStates := snapTaskStatusForChange(chg)
+	// The pi-kernel failed in error-trigger and got rolled back
+	c.Check(taskStates["pi-kernel:error-trigger"], Equals, state.ErrorStatus)
+	c.Check(taskStates["pi-kernel:mount-snap"], Equals, state.UndoneStatus)
+	// But the pi gadget was installed just fine
+	c.Check(taskStates["pi:download-snap"], Equals, state.DoneStatus)
+	c.Check(taskStates["pi:link-snap"], Equals, state.DoneStatus)
+
+	// Note that the undo of the kernel did *not* revert the DTBs on
+	// disk. The reasons is that we never undo asset updates on the
+	// basis that if the system booted they are probably good enough.
+	// A really broken DTB can brick the device, the new DTB is written
+	// to disk, the system reboot and neither new kernel nor fallback
+	// kernel will boot because there is no A/B DTB. This is a flaw
+	// of the PI and u-boot.
```

Comment:
```suggestion
	// A really broken DTB can brick the device if the new DTB is written
	// to disk, the system reboots and neither new kernel nor fallback
	// kernel will boot because there is no A/B DTB. This is a flaw
	// of the Pi and u-boot.
```